usaspending_api/common/filters/time_period/time_period.py

```diff
@@ -31,7 +32,7 @@ def start_date(self):
         pass
 
     @abstractmethod
-    def end_date(self):
+    def end_date(self) -> str:
```

Comment:
This confuses me that the `end_date` and `start_date` would return a string. It leaves a lot open to interpretation of what date format that string may need to be in, and different filtering implementations may need to actually provide the date filter string in _different_ formats according to the data they're filtering against. So it may be best to make this return the actual `datetime` object, and let it be formatted as needed at the point where the filter is applied. 

If it can't be done as a date object, then the type hints do help a lot, but you should also add some cautions and caveats in the docs for this method as to how and when the date should be converted to a string, and how to determine what string format it should be in.

Reply:
It doesn't return a datetime. It returns the value in the filter which is a string

Reply:
Maybe I need to change the name of this method then?

Reply:
Right, but my comment was indicating it probably should return a datetime -- or caution in the docstring about how best turn a datetime into a string. 

I think what may be confusing is that we're passing in a filter to this filter's constructor in some of the concrete impls? That seems convoluted, and I was thinking this would be the origin of building-up a filter from primitive values (i.e. datetime objects).

Reply:
Building up time period filters does not involve any actual date objects. It's all strings, that's what our exiting ES_Q expect. We are passing the time period input from the user into this filter's constructor. I think the name of the methods need to just be updated.

Reply:
I don't think we should return a datetime because calling code doesn't need it as a date object. It doesn't need that anywhere. Which is why I suggested changing the method name.

Reply:
Maybe what we need is a concrete object (not a dictionary) representing the Filter object that is coming in from the API request? 

Or a dataclass. 

Then 
- Api request -> JSON -> <convert from json reqeust filter to python object> = `<<RequestFilterObject>>`
- `ConcreteTimePeriod(request_date_filter: RequestFilterObject)`
- and decide if RequestFilterObject is always used, and put up on AbstractTimePeriod's constructor,
    - Or if TimePeriod has an explicitly constructor with decomposing params into of `TimePeriod(start, end, start_range, end_range)` etc.
    - and concrete impls like `ConcreteTimePeriod(request_date_filter: RequestFilterObject)` then just decompose that RequestFilterObject input param and make a call up to the explicit `TimePeriod(start, end, start_range, end_range)`  constructor with all the parts it extracted out of the  RequestFilterObject

Reply:
Wouldn't factory pattern make more sense for choosing which concrete implementation to use? We already know which concrete implementation to use. For example, see usaspending_api/search/v2/views/spending_over_time.py the method 182

Reply:
So now instead of having a method that has a bunch of `if` statements directing flow based on whether the query type is awards or transactions, the method works with any subsclass of a `AbstractTimePeriod`. This subclass encapsulates it's custom behavior.

Reply:
> I feel like we are over thinking it and a factory may be overkill. Transaction searches use time period consistently and differently than other apps. We know whether we want to use a transaction search time period the moment a request hits our api it doesn't change in the middle of a request.

This may mean that this filter translation is in fact coupled to the API and needs to live closer to the API code rather than a globally reusable and generic time period filter. 

Also, is "TransactionSearch" here indicating
- a) the `TransactionSearch` ORM object which is tied to the `transaction_search` DB table?
- b) or, the transaction data stored in the `*transactions` elasticsearch indexes?
- c) or, the user activity of "using Advanced Search to search across transaction data"

I feel like there's an implication in the name that I may be misinterpreting. It leads me to believe `a`

Reply:
These classes are intended to solely represent the time period object that our api accepts. It's coupled to that concept.

Reply:
Moving them to awards doesn't make sense, these are not awards specific. Any ideas on where else they could be moved to?

Reply:
Moving them to awards doesn't make sense, these are not awards specific. Any ideas on where else they could be moved to?

Reply:
Going move the classes